### PR TITLE
Numba

### PR DESCRIPTION
**Pull request by harpolea**
This adds `numba` versions of all the fortran interface functions. (Non-exhaustive) profiling indicates that the `numba` versions are about 5x slower than the fortran versions, but 20-100x faster than plain python. For most problems, the `numba` versions should be sufficient as the overall time will still be dominated by the plain python code.

The solutions for both versions agree for all but the incompressible and low Mach solvers - I think this is to be expected given they both use the multigrid solver and so incur more numerical error overall? I couldn't find a way to pass in an error parameter to `nbval`, so for CI purposes I have made two versions of `solver-test.ipynb`. This isn't ideal - it might be better to split off the tests that use the MG solver into a separate notebook and e.g. sanitize the output so nbval only checks up to 3sf.

I've changed `mk.sh` so it builds the `numba` versions by default, but if you run it with the argument `fortran` it will build the fortran functions instead. You can revert back to the `numba` versions again by running `./mk.sh clean`.

### CODE REVIEW

File: advection_fv4/interface.py [hunk 1]
```diff
@@ -0,0 +1,302 @@
...
+
+    C2 = 1.25
+    C3 = 0.1
+
+    nx = qx - 2 * ng
+    ny = qy - 2 * ng
+    ilo = ng
+    ihi = ng + nx
+    jlo = ng
+    jhi = ng + ny
```
Comment (zingale) [inline advection_fv4/interface.py]: can we get away with just passing in the grid object now?  then we would not need to recompute these

File: advection_fv4/interface.py [hunk 2]
```diff
@@ -0,0 +1,302 @@
...
+
+    # we need interface values on all faces of the domain
+    if (idir == 1):
+
+        for j in range(jlo - 1, jhi + 1):
```
Comment (zingale) [inline advection_fv4/interface.py]: shouldn't this be `jhi+2`?

File: advection_fv4/interface.py [hunk 3]
```diff
@@ -0,0 +1,302 @@
...
+            for i in range(ilo - 2, ihi + 3):
```
Comment (zingale) [inline advection_fv4/interface.py]: `ihi+3`?

Comment (zingale) [reply advection_fv4/interface.py]: there are a lot of these.  I think range is up to, but not including the upper limit, right?
oh... wait... is it because we are defining ihi and jhi differently now?

File: advection_fv4/interface.py [hunk 4]
```diff
@@ -0,0 +1,302 @@
...
+
+                # limit? MC Eq. 24 and 25
+                if (dafm[i, j] * dafp[i, j] <= 0.0 or
+                        (a[i, j] - a[i - 2, j]) * (a[i + 2, j] - a[i, j]) <= 0.0):
+
+                    # we are at an extrema
+
+                    s = np.copysign(1.0, d2ac[i, j])
+                    if (s == np.copysign(1.0, d2ac[i - 1, j]) and s == np.copysign(1.0, d2ac[i + 1, j]) and
+                            s == np.copysign(1.0, d2af[i, j])):
```
Comment (zingale) [inline advection_fv4/interface.py]: copysign? or sign?

File: .travis.yml [hunk 5]
```diff
@@ -19,11 +28,11 @@ install:
 before_script:
   - export PYTHONPATH=$PYTHONPATH:$(pwd)
   - export PYRO_HOME=$(pwd)
-  - ./mk.sh
+  - ./mk.sh $MK_ARGS
 
 script:
-  - flake8 .
-  - pytest -v --cov=. --cov-config .coveragerc --nbval --ignore=docs --ignore=./multigrid/variable_coeff_elliptic.ipynb --ignore=examples/mesh --ignore=examples/multigrid
+  - flake8 --ignore=W504,F403,E226,E126,E127,E128,E129,E241,E741 .
```
Comment (zingale) [inline .travis.yml]: there already is a `.flake8`, so maybe we can add the ignores there?

File: compressible/interface_f.f90 [hunk 6]
```diff
@@ -296,6 +296,8 @@ subroutine riemann_cgf(idir, qx, qy, ng, &
            ut_l    = U_l(i,j,ixmom)/rho_l
         endif
 
+        rhoe_l = U_l(i,j,iener) - 0.5*rho_r*(un_l**2 + ut_l**2)
+
```
Comment (zingale) [inline compressible/interface_f.f90]: why is this needed?

**Comment by harpolea** [comment]
If we're happy with the `numba` performance, then I agree it will be much easier to just maintain one version so it would make sense to ditch the fortran

File: advection_fv4/interface.py (hunk 1 above)
Comment (harpolea) [reply advection_fv4/interface.py]: We could - I mainly left them in so that the python and fortran functions looked as similar as possible and had the same inputs

File: advection_fv4/interface.py (hunk 2 above)
Comment (harpolea) [reply advection_fv4/interface.py]: I don't think so - I changed `ihi` and `jhi` in the python version so that we could keep the nice symmetric iteration limits

File: advection_fv4/interface.py (hunk 3 above)
Comment (harpolea) [reply advection_fv4/interface.py]: Yes, I added 1 to `ihi` and `jhi` in the python version

File: advection_fv4/interface.py (hunk 4 above)
Comment (harpolea) [reply advection_fv4/interface.py]: [`np.copysign`](https://docs.scipy.org/doc/numpy-1.15.1/reference/generated/numpy.copysign.html) seems to have the same behaviour as Fortran's [`sign`](https://gcc.gnu.org/onlinedocs/gfortran/SIGN.html)

File: compressible/interface_f.f90 (hunk 6 above)
Comment (harpolea) [reply compressible/interface_f.f90]: I noticed when I did the python version that `rhoe_l` is not calculated before that point then is used to calculate `p_l`. The `rho_r` should be a `rho_l` though.

File: .travis.yml (hunk 5 above)
Comment (harpolea) [reply .travis.yml]: Fixed by 69dec2946

File: compressible/interface_f.f90 (hunk 6 above)
Comment (zingale) [reply compressible/interface_f.f90]: huh... indeed.  How did it even work then?

File: advection_fv4/interface.py (hunk 1 above)
Comment (zingale) [reply advection_fv4/interface.py]: okay.  Perhaps then after the merge and once we eliminate the Fortran stuff, we can simplify things then.

**Comment by zingale** [comment]
it seems some of the tests, like compressible / rt take more steps now, and don't have a corresponding benchmark.

**Comment by harpolea** [comment]
Oh that's interesting given that the Numba code is so similar to the Fortran - I wouldn't have expected the results to be so different that the number of timesteps changed.

**Comment by zingale** [comment]
tests seem to agree to roundoff.  Still get 4th order convergence for the `compressible_fv4` solver.
